sql/parser/type_check.go

```diff
@@ -0,0 +1,388 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+package parser
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/util"
+)
+
+// TypeCheckExpr type-checks an SQL expression. Expression checking is
+// a mostly straightforward walk over the parse tree.
+// TODO(tamird): tuple handling is currently crappy.
+func TypeCheckExpr(expr Expr) (reflect.Type, error) {
```

Comment:
My comment about using `Datum` to hold type information instead of `reflect.Type` seems to have gotten lost. If you returned a `Datum` here then you would be able to distinguish between different types of tuples.